File: microservices/authentication/adapters/inbound/rabbitmq/rabbitmq.inbound-adapter.go
package rabbitMQInboundAdapter

import (
	"log"

	"github.com/streadway/amqp"
	"google.golang.org/protobuf/proto"

	"authentication/generated/proto/messages"
	"authentication/ports"
	"authentication/types"
	"authentication/utils"
)

type RabbitMQInboundAdapter struct {
	ApplicationLayer ports.ApplicationPort

	Channel *amqp.Channel
}

func(instance *RabbitMQInboundAdapter) CreateChannel(connection *amqp.Connection) {
	var error error

	instance.Channel, error= connection.Channel( )
	if error != nil {
		log.Fatal("💀 error creating rabbitMQ channel : ", error.Error( ))}

	utils.DeclareRabbitMQQueue(instance.Channel, utils.AuthenticationQueueName)

	log.Println("🔥 created rabbitMQ channel for message consumption")
}

func(instance *RabbitMQInboundAdapter) DestroyChannel( ) {
	if instance.Channel != nil {
		instance.Channel.Close( )}
}

func(instance *RabbitMQInboundAdapter) StartMessageConsumption( ) {

	newMessages, error := instance.Channel.Consume(utils.AuthenticationQueueName, "", false, false, false, false, nil)
	if error != nil {
		log.Fatalf("💀 error consuming from queue %s : %s", utils.AuthenticationQueueName, error.Error( ))}

	for message := range newMessages {
		var unmarshalledMessage messages.Message

		error := proto.Unmarshal(message.Body, &unmarshalledMessage)
		if error != nil {
			log.Fatalf("💀 error unmarshalling new message received from queue : %s", error.Error( ))}

		switch unmarshalledMessage.MessageType {

			case utils.RegisterUser_MessageType:
				var request messages.RegisterUserIncomingMessage

				error := proto.Unmarshal(message.Body, &request)
				if error != nil {
					log.Fatalf("💀 error unmarshalling `set temporary user verified` request : %s", error.Error( ))

					message.Ack(false)
					continue
				}

				output := instance.ApplicationLayer.Register(
					&types.RegisterParameters{
						Email: request.Email,
					})

				if output.Error != nil {
					message.Ack(false); continue}

				message.Ack(true)

			default:
				log.Printf("unknown message of type %s received", unmarshalledMessage.MessageType)
		}
	}
}

func CreateRabbitMQInboundAdapter(applicationLayer ports.ApplicationPort) *RabbitMQInboundAdapter {
	return &RabbitMQInboundAdapter{
		ApplicationLayer: applicationLayer,
	}
}

File: microservices/otp/src/main/java/otp/adapters/inbound/rabbitmq/RabbitMQInboundAdapter.java
package otp.adapters.inbound.rabbitmq;

import java.io.Closeable;
import java.io.IOException;
import java.util.Collections;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicReference;

import com.google.protobuf.InvalidProtocolBufferException;
import com.otp.generated.proto.Message;
import com.otp.generated.proto.SendOTPIncomingMessage;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.DeliverCallback;

import otp.exceptions.AppStartupException;
import otp.ports.inbound.APIPort;
import otp.types.SendOTPOutput;
import otp.types.SendOTPParameters;
import otp.utils.MessageQueues;

public class RabbitMQInboundAdapter implements Closeable {
    Channel channel;
    Thread workerThread;

    APIPort applicationLayer;

    public RabbitMQInboundAdapter(Connection connection, APIPort applicationLayer) throws AppStartupException {
        try {
            this.channel= connection.createChannel( );
            this.channel.queueDeclare(MessageQueues.otp.toString( ), false, false, false, Collections.emptyMap( ));
        }
            catch(IOException error) {
                throw new AppStartupException("error creating rabbitMQ consumption channel : " + error.getMessage( ));}

        this.applicationLayer= applicationLayer;
    }

    DeliverCallback processMessage= (consumerTag, message) -> {
        Boolean acknowledge= true;

        try {
            Message genericRequest= Message.parser( ).parseFrom(message.getBody( ));

            // TODO: use values from `MessageTypes` enum instead for matching
            switch(genericRequest.getMessageType( )) {

                case "SendOTP":
                    var request= SendOTPIncomingMessage.parser( ).parseFrom(message.getBody( ));

                    SendOTPOutput result=
                        this.applicationLayer.sendOTP(new SendOTPParameters(
                            request.getEmail( )
                        ));

                    acknowledge= result.isEmailSent( );

                    break;

                default:
                    System.out.println("unknown type of message received from rabbitMQ");
                    acknowledge= true;
            }
        }
            catch(InvalidProtocolBufferException exception) {
                System.out.println("error parsing rabbitMQ message");}

        this.channel.basicAck(message.getEnvelope( ).getDeliveryTag( ), acknowledge);
    };

    public void consumeMessages( ) throws AppStartupException {
        AtomicReference<IOException> errorInWorkerThread= new AtomicReference<IOException>(null);

        this.workerThread= new Thread(
            ( ) -> {
                while(this.workerThread.isInterrupted( )) {
                    try {
                        this.channel.basicConsume(MessageQueues.otp.toString( ), false, processMessage, consumerTag -> { });
                    }
                        catch(IOException error) { errorInWorkerThread.set(error); }
                }
            }
        );

        if(errorInWorkerThread != null)
            throw new AppStartupException("error trying to consume messages from `otp` queue : " + errorInWorkerThread.get( ).getMessage( ));
    }

    @Override
    public void close( ) {
        try {
            if(this.channel != null)
                this.channel.close( );

        } catch(IOException | TimeoutException exception) { }
    }

}